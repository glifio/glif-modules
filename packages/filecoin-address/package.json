{
  "name": "@glif/filecoin-address",
  "version": "1.0.0-beta.5",
  "description": "The Filecoin address type",
  "main": "./dist/index.js",
  "browser": "./lib/index.js",
  "module": "./module/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "clean": "../../node_modules/.bin/rimraf dist/ lib/ module/",
    "build-browser": "../../node_modules/.bin/cross-env BABEL_ENV=browser babel ./src --out-dir ./lib --extensions \".ts,.tsx\" --source-maps --ignore '**/__tests__'",
    "build-module": "cross-env BABEL_ENV=module babel ./src --out-dir ./module --extensions \".ts,.tsx\" --source-maps --ignore '**/__tests__'",
    "build-node": "babel ./src/ --out-dir ./dist/ --extensions \".ts,.tsx\" --source-maps --ignore '**/__tests__'",
    "build-types": "../../node_modules/.bin/tsc --emitDeclarationOnly",
    "build": "npm run clean && npm run build-types && npm run build-node && npm run build-browser && npm run build-module",
    "prepublishOnly": "npm run build",
    "test": "npm run build && jest",
    "test:watch": "npm run build && jest --watch",
    "type-check": "../../node_modules/.bin/tsc --noEmit",
    "type-check:watch": "npm run type-check -- --watch"
  },
  "author": "Infinite Scroll <squad@infinitescroll.org> (https://infinitescroll.org)",
  "license": "(Apache-2.0 OR MIT)",
  "dependencies": {
    "blakejs": "1.1.0",
    "borc": "2.1.1",
    "leb128": "0.0.5",
    "uint8arrays": "^1.1.0"
  },
  "devDependencies": {},
  "files": [
    "dist",
    "lib",
    "module"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/glifio/modules/tree/primary/packages/filecoin-address"
  }
}
