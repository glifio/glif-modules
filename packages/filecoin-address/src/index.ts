import * as leb from 'leb128'
import { blake2b } from 'blakejs'
import { base32 as base32Function } from './base32'
import * as uint8arrays from 'uint8arrays'
import { Protocol } from './protocol'
import { Network } from './network'

export * from './network'
export * from './protocol'

const base32 = base32Function('abcdefghijklmnopqrstuvwxyz234567')

export class Address {
  readonly str: Uint8Array
  readonly _protocol: Protocol

  constructor(str: Uint8Array) {
    if (!str || str.length < 1) throw new Error('Missing str in address')
    this.str = str
    this._protocol = this.str[0] as Protocol
    if (!Protocol[this._protocol]) {
      throw new Error(`Invalid protocol ${this._protocol}`)
    }
  }

  protocol(): Protocol {
    return this._protocol
  }

  payload(): Uint8Array {
    return this.str.slice(1, this.str.length)
  }
}

export function bigintToArray(v: string | BigInt | number): Uint8Array {
  let tmp = BigInt(v).toString(16)
  if (tmp.length % 2 === 1) tmp = `0${tmp}`
  return uint8arrays.fromString(tmp, 'base16')
}

export function getChecksum(ingest: string | Uint8Array): Uint8Array {
  return blake2b(ingest, null, 4)
}

export function validateChecksum(
  ingest: string | Uint8Array,
  expect: Uint8Array
) {
  const digest = getChecksum(ingest)
  return Buffer.compare(Buffer.from(digest), expect)
}

export function newAddress(protocol: Protocol, payload: Uint8Array): Address {
  const protocolByte = new Uint8Array([protocol])
  return new Address(Buffer.concat([protocolByte, payload]))
}

export function decode(address: string): Address {
  checkAddressString(address)

  const network = address.slice(0, 1) as Network
  const protocol = parseInt(address.slice(1, 2)) as Protocol
  const raw = address.substring(2, address.length)
  const protocolByte = new Uint8Array([protocol])

  if (protocol === Protocol.ID) {
    return newAddress(protocol, Buffer.from(leb.unsigned.encode(raw)))
  }

  const payloadChecksum = base32.decode(raw)
  const length = payloadChecksum.length
  const payload = payloadChecksum.slice(0, length - 4)
  const checksum = payloadChecksum.slice(length - 4, length)
  if (validateChecksum(uint8arrays.concat([protocolByte, payload]), checksum)) {
    throw Error("Checksums don't match")
  }

  const addressObj = newAddress(protocol, payload)
  if (encode(network, addressObj) !== address)
    throw Error(`Did not encode this address properly: ${address}`)

  return addressObj
}

export function encode(network: string, address: Address): string {
  if (!address || !address.str) throw Error('Invalid address')
  const payload = address.payload()

  switch (address.protocol()) {
    case 0: {
      return (
        network +
        String(address.protocol()) +
        leb.unsigned.decode(address.payload())
      )
    }
    default: {
      const protocolByte = new Uint8Array([address.protocol()])
      const checksum = getChecksum(Buffer.concat([protocolByte, payload]))
      const bytes = Buffer.concat([payload, Buffer.from(checksum)])
      return String(network) + String(address.protocol()) + base32.encode(bytes)
    }
  }
}

export function newFromString(address: string): Address {
  return decode(address)
}

export function validateAddressString(string: string): boolean {
  try {
    checkAddressString(string)
    return true
  } catch (error) {
    return false
  }
}

export function checkAddressString(address: string) {
  if (!address) throw Error('No bytes to validate.')
  if (address.length < 3) throw Error('Address is too short to validate.')
  if (address[0] !== 'f' && address[0] !== 't') {
    throw Error('Unknown address network.')
  }

  const protocol = parseInt(address[1]) as Protocol
  switch (protocol) {
    case Protocol.ID: {
      if (address.length > 22) throw Error('Invalid ID address length.')
      break
    }
    case Protocol.SECP256K1: {
      if (address.length !== 41)
        throw Error('Invalid secp256k1 address length.')
      break
    }
    case Protocol.ACTOR: {
      if (address.length !== 41) throw Error('Invalid Actor address length.')
      break
    }
    case Protocol.BLS: {
      if (address.length !== 86) throw Error('Invalid BLS address length.')
      break
    }
    default: {
      throw new Error('Invalid address protocol.')
    }
  }
}

export default {
  Address,
  newAddress,
  newFromString,
  bigintToArray,
  decode,
  encode,
  getChecksum,
  validateChecksum,
  validateAddressString,
  checkAddressString,
  Network,
  Protocol
}
