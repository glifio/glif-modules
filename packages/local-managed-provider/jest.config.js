module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'node',
  collectCoverageFrom: ['**/*.{ts,js}', '!**/node_modules/**'],
  testPathIgnorePatterns: ['/node_modules/'],
  moduleNameMapper: {
    "^@zondax/filecoin-signing-tools$": "@zondax/filecoin-signing-tools/nodejs"
  },
  testTimeout: 50000,
}
