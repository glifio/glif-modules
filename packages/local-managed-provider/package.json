{
  "name": "@glif/local-managed-provider",
  "version": "1.0.0-beta.5",
  "description": "a javascript package to send filecoin to addresses",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "clean": "../../node_modules/.bin/rimraf dist/",
    "build": "npm run clean && tsc --project tsconfig.json",
    "lint": "tslint -p tsconfig.json",
    "prepublishOnly": "npm run build",
    "test:build": "npm run build && jest --forceExit",
    "test": "npm run build && jest --forceExit",
    "test:watch": "npm run build && jest --watch"
  },
  "author": "Infinite Scroll <squad@infinitescroll.org> (https://infinitescroll.org)",
  "license": "(Apache-2.0 OR MIT)",
  "dependencies": {
    "@glif/filecoin-message": "^1.0.0-beta.5",
    "@glif/filecoin-address": "^1.0.0-beta.5",
    "@glif/filecoin-wallet-provider": "^1.0.0-beta.5",
    "@zondax/filecoin-signing-tools": "^0.12.0",
    "bignumber.js": "^9.0.1",
    "uint8arrays": "^1.1.0"
  },
  "devDependencies": {
    "cids": "^1.0.0",
    "cross-env": "7.0.2",
    "eslint": "6.1.0",
    "eslint-config-airbnb": "18.1.0",
    "eslint-config-prettier": "6.7.0",
    "eslint-plugin-import": "2.20.2",
    "jest": "26.6.2",
    "prettier": "^1.19.1",
    "tslint": "^6.1.3",
    "tslint-config-prettier": "^1.18.0",
    "typescript": "^4.0.2",
    "ts-jest": "^26.4.3",
    "@types/jest": "^26.0.15"
  },
  "files": [
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/glifio/modules/tree/primary/packages/local-managed-provider"
  },
  "gitHead": "27ff58865a447fe70913209a04756cf5ca67d3df"
}
