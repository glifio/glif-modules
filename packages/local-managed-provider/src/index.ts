export * from './local-managed-provider'
