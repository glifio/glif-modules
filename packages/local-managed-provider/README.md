# Local managed provider

:warning: Active development. Unstable. Breaking Changes. You get the point.

This wallet subprovider handles signing by a managed raw private key. Supports secp256k1 and BLS keys.

## Installation

```
npm add @glif/local-managed-provider @glif/filecoin-address @glif/filecoin-message
```

## Usage

```js
import { LocalManagedProvider } from '@glif/local-managed-provider'
import { Network } from '@glif/filecoin-address'
import { SignedLotusMessage } from '@glif/filecoin-message'
// Hex of JSON, same as returned by Lotus `wallet export` command.
const secp256k1Key =
  '7b2254797065223a22736563703235366b31222c22507269766174654b6579223a2257587362654d5176487a366f5668344b637262633045642b31362b3150766a6a504f3753514931355031343d227d'
const provider = new LocalManagedProvider(secp256k1Key, Network.TEST)
const accounts = await provider.getAccounts() // t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka
const sig: SignedLotusMessage = await provider.sign(address, {
  From: address,
  To: address,
  Value: '0',
  Method: 0,
  GasPrice: '1',
  GasFeeCap: '1',
  GasPremium: '1',
  GasLimit: 1000,
  Nonce: 0,
  Params: '',
}) // Returns SignedLotusMessage
```
