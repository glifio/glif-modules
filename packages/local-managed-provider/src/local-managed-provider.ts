import { WalletSubProvider } from '@glif/filecoin-wallet-provider'
import { LotusMessage, SignedLotusMessage } from '@glif/filecoin-message'
import { Network } from '@glif/filecoin-address'
import {
  privateKeyContainer,
  PrivateKeyContainer,
} from './private-key-container'

export class LocalManagedProvider implements WalletSubProvider {
  private readonly privateKey: PrivateKeyContainer

  constructor(privateKey: string, network: Network) {
    this.privateKey = privateKeyContainer(privateKey, network)
  }

  async getAccounts(): Promise<string[]> {
    return [this.privateKey.address]
  }

  async sign(from: string, message: LotusMessage): Promise<SignedLotusMessage> {
    if (from === this.privateKey.address) {
      return this.privateKey.sign(message)
    } else {
      throw new Error(`Can only sign with address ${this.privateKey.address}`)
    }
  }
}
