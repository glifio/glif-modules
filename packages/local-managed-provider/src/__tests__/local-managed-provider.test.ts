import { SingleKeyProvider } from '../local-managed-provider'
import { CoinType } from '@glif/filecoin-address'

const secp256k1Key =
  '7b2254797065223a22736563703235366b31222c22507269766174654b6579223a2257587362654d5176487a366f5668344b637262633045642b31362b3150766a6a504f3753514931355031343d227d'

const blsKey =
  '7b2254797065223a22626c73222c22507269766174654b6579223a226e586841424f4163796856504b48326155596261796f4475752f4c6f32515a2b6662622f6f736a2f34456f3d227d'

describe('secp256k1', () => {
  const provider = new SingleKeyProvider(secp256k1Key)
  test('#getAccounts', async () => {
    const accounts = await provider.getAccounts(0, 0, CoinType.TEST)
    expect(accounts).toEqual(['t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka'])
  })
  test('#keyDerive', async () => {
    expect(await provider.keyDerive('')).toBe(secp256k1Key)
  })
  test('#sign', async () => {
    const accounts = await provider.getAccounts(0, 0, CoinType.TEST)
    const address = accounts[0]
    const sig = await provider.sign(address, {
      From: address,
      To: address,
      Value: '0',
      Method: 0,
      GasFeeCap: '1',
      GasPremium: '1',
      GasLimit: 1000,
      Nonce: 0,
      Params: '',
    })
    expect(sig).toMatchSnapshot()
  })
})

describe('bls', () => {
  const provider = new SingleKeyProvider(blsKey)
  test('#getAccounts', async () => {
    const accounts = await provider.getAccounts(0, 0, CoinType.TEST)
    expect(accounts).toEqual([
      't3vbrwhphivdxyvs3pxpp54w73664bgxmb7ed4du4ohhu3dc6f5y264cs72yluw7mjbwnlrtzq543ys57plzka',
    ])
  })
  test('#keyDerive', async () => {
    expect(await provider.keyDerive('')).toBe(blsKey)
  })
  test('#sign', async () => {
    const accounts = await provider.getAccounts(0, 0, CoinType.TEST)
    const address = accounts[0]
    const sig = await provider.sign(address, {
      From: address,
      To: address,
      Value: '0',
      Method: 0,
      GasFeeCap: '1',
      GasPremium: '1',
      GasLimit: 1000,
      Nonce: 0,
      Params: '',
    })
    expect(sig).toMatchSnapshot()
  })
})
