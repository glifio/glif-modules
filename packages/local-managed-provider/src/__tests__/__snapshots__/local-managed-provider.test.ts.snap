// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bls #sign 1`] = `
{
  "Message": {
    "From": "t3vbrwhphivdxyvs3pxpp54w73664bgxmb7ed4du4ohhu3dc6f5y264cs72yluw7mjbwnlrtzq543ys57plzka",
    "GasFeeCap": "1",
    "GasLimit": 1000,
    "GasPremium": "1",
    "Method": 0,
    "Nonce": 0,
    "Params": "",
    "To": "t3vbrwhphivdxyvs3pxpp54w73664bgxmb7ed4du4ohhu3dc6f5y264cs72yluw7mjbwnlrtzq543ys57plzka",
    "Value": "0",
  },
  "Signature": {
    "Data": "gTjLEAanvWII9SOrRpdgEIV3BNZvEmP1/QocIlJA3eo9WmUT28LLkMq4gYBI8aggEzYw0XhRKCFKY191V2EM16amLiwkmcs+qAJJcXAAoO+Am2enYxFHhUWcGyRILHYc",
    "Type": 2,
  },
}
`;

exports[`secp256k1 #sign 1`] = `
{
  "Message": {
    "From": "t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka",
    "GasFeeCap": "1",
    "GasLimit": 1000,
    "GasPremium": "1",
    "Method": 0,
    "Nonce": 0,
    "Params": "",
    "To": "t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka",
    "Value": "0",
  },
  "Signature": {
    "Data": "l3WF6gkTMr1ZD8YUT186/0I7Z+ftoUWs7/2J/vt86qMD5u582Oh+gEcBUq+OaZCNfYdsvq/MTJn8aU07PquT/wE=",
    "Type": 1,
  },
}
`;
